src/Atmos/Model/bc_momentum.jl

```diff
@@ -0,0 +1,175 @@
+"""
+struct FreeSlip <: MomentumDragBC end
+
+
+
+function atmos_momentum_boundary_state!(
+    nf::NumericalFluxNonDiffusive,
+    bc_momentum::Impenetrable{FreeSlip},
+    atmos,
+    state⁺,
+    aux⁺,
+    n,
+    state⁻,
+    aux⁻,
+    bctype,
+    t,
+    args...,
+)
+    state⁺.ρu -= 2 * dot(state⁻.ρu, n) .* SVector(n)
+end
+function atmos_momentum_boundary_state!(
+    nf::NumericalFluxGradient,
+    bc_momentum::Impenetrable{FreeSlip},
+    atmos,
+    state⁺,
+    aux⁺,
+    n,
+    state⁻,
+    aux⁻,
+    bctype,
+    t,
+    args...,
+)
+    state⁺.ρu -= dot(state⁻.ρu, n) .* SVector(n)
+end
+function atmos_momentum_normal_boundary_flux_diffusive!(
+    nf,
+    bc_momentum::Impenetrable{FreeSlip},
+    atmos,
+    args...,
+) end
+
+
+
+"""
+    NoSlip() :: MomentumDragBC
+
+Zero momentum at the boundary.
+"""
+struct NoSlip <: MomentumDragBC end
+
+function atmos_momentum_boundary_state!(
+    nf::NumericalFluxNonDiffusive,
+    bc_momentum::Impenetrable{NoSlip},
+    atmos,
+    state⁺,
+    aux⁺,
+    n,
+    state⁻,
+    aux⁻,
+    bctype,
+    t,
+    args...,
+)
+    state⁺.ρu = -state⁻.ρu
+end
+function atmos_momentum_boundary_state!(
+    nf::NumericalFluxGradient,
+    bc_momentum::Impenetrable{NoSlip},
+    atmos,
+    state⁺,
+    aux⁺,
+    n,
+    state⁻,
+    aux⁻,
+    bctype,
+    t,
+    args...,
+)
+    state⁺.ρu = zero(state⁺.ρu)
+end
+function atmos_momentum_normal_boundary_flux_diffusive!(
+    nf,
+    bc_momentum::Impenetrable{NoSlip},
+    atmos,
+    args...,
+) end
+
+
+"""
+    DragLaw(fn) :: MomentumDragBC
+
+Drag law for momentum parallel to the boundary. The drag coefficient is
+`C = fn(state, aux, t, normPu1⁻)`, where `normPu1⁻` is the internal speed
+parallel to the boundary.
+"""
+struct DragLaw{FN} <: MomentumDragBC
+    fn::FN
+end
+function atmos_momentum_boundary_state!(
+    nf::Union{NumericalFluxNonDiffusive, NumericalFluxGradient},
+    bc_momentum::Impenetrable{DL},
+    atmos,
+    state⁺,
+    aux⁺,
+    n,
+    state⁻,
+    aux⁻,
+    bctype,
+    t,
+    args...,
+) where {DL <: DragLaw}
+    atmos_momentum_boundary_state!(
+        nf,
+        Impenetrable(FreeSlip()),
+        atmos,
+        state⁺,
+        aux⁺,
+        n,
+        state⁻,
+        aux⁻,
+        bctype,
+        t,
+        args...,
+    )
+end
+function atmos_momentum_normal_boundary_flux_diffusive!(
+    nf,
+    bc_momentum::Impenetrable{DL},
+    atmos,
+    fluxᵀn,
+    n,
+    state⁻,
+    diff⁻,
+    hyperdiff⁻,
+    aux⁻,
+    state⁺,
+    diff⁺,
+    hyperdiff⁺,
+    aux⁺,
+    bctype,
+    t,
+    state1⁻,
+    diff1⁻,
+    aux1⁻,
+) where {DL <: DragLaw}
+
+    u1⁻ = state1⁻.ρu / state1⁻.ρ
+    Pu1⁻ = u1⁻ - dot(u1⁻, n) .* SVector(n)
+    normPu1⁻ = norm(Pu1⁻)
+    # NOTE: difference from design docs since normal points outwards
+    C = bc_momentum.drag.fn(state⁻, aux⁻, t, normPu1⁻)
+    τn = C * normPu1⁻ * Pu1⁻
+
+    fluxᵀn.ρu += state⁻.ρ * τn
+    fluxᵀn.ρe += state⁻.ρu' * τn
```

Comment:
Expression is OK - same comment as in energy b.c. file regarding having += / -= be consistent across files / functions? (Makes it easier to interpret I think)

Reply:
Also notation / clarity - `P` implying the horizontal projection.

Reply:
they both involve projections of normals, so signs here are consistent.